File: src/batch_inference.cpp
#include "inference.h"
#include <iostream>
#include <iomanip>
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <chrono>
#include <filesystem>
#include <numeric>

namespace fs = std::filesystem;

// ANSI color codes for terminal output
#define RESET   "\033[0m"
#define RED     "\033[1;31m"
#define GREEN   "\033[1;32m"
#define YELLOW  "\033[1;33m"
#define BLUE    "\033[1;34m"
#define MAGENTA "\033[1;35m"
#define CYAN    "\033[1;36m"

struct Statistics {
    int total = 0;
    int correct = 0;
    std::vector<std::vector<int>> confusion_matrix;
    std::map<int, int> per_class_total;
    std::map<int, int> per_class_correct;
    std::vector<double> inference_times;

    Statistics() {
        // Initialize 10x10 confusion matrix
        confusion_matrix.resize(10, std::vector<int>(10, 0));
        for (int i = 0; i < 10; ++i) {
            per_class_total[i] = 0;
            per_class_correct[i] = 0;
        }
    }

    void addPrediction(int true_label, int predicted, double inference_time) {
        total++;
        per_class_total[true_label]++;
        inference_times.push_back(inference_time);

        confusion_matrix[true_label][predicted]++;

        if (true_label == predicted) {
            correct++;
            per_class_correct[true_label]++;
        }
    }

    double getOverallAccuracy() const {
        return total > 0 ? (100.0 * correct / total) : 0.0;
    }

    double getClassAccuracy(int digit) const {
        int class_total = per_class_total.at(digit);
        int class_correct = per_class_correct.at(digit);
        return class_total > 0 ? (100.0 * class_correct / class_total) : 0.0;
    }

    double getAverageInferenceTime() const {
        if (inference_times.empty()) return 0.0;
        double sum = std::accumulate(inference_times.begin(), inference_times.end(), 0.0);
        return sum / inference_times.size();
    }
};

void printProgressBar(int current, int total, int true_label) {
    const int bar_width = 50;
    float progress = static_cast<float>(current) / total;
    int pos = static_cast<int>(bar_width * progress);

    std::cout << "\r  Digit " << true_label << ": [";
    for (int i = 0; i < bar_width; ++i) {
        if (i < pos) std::cout << "█";
        else std::cout << "░";
    }
    std::cout << "] " << current << "/" << total << " images" << std::flush;
}

void printHeader() {
    std::cout << "\n╔" << std::string(78, '=') << "╗\n";
    std::cout << "║" << std::string(20, ' ') << "C++ BATCH INFERENCE REPORT"
              << std::string(32, ' ') << "║\n";
    std::cout << "╚" << std::string(78, '=') << "╝\n\n";
}

void printOverallStats(const Statistics& stats) {
    std::cout << "📊 Overall Statistics:\n";
    std::cout << "   Total images:     " << stats.total << "\n";
    std::cout << "   Correct:          " << stats.correct << "\n";
    std::cout << "   Incorrect:        " << (stats.total - stats.correct) << "\n";
    std::cout << "   Accuracy:         " << std::fixed << std::setprecision(2)
              << stats.getOverallAccuracy() << "%\n";
}

void printPerformance(const Statistics& stats) {
    double avg_time = stats.getAverageInferenceTime() * 1000.0; // Convert to ms
    std::cout << "\n⚡ Performance:\n";
    std::cout << "   Avg inference:    " << std::fixed << std::setprecision(2)
              << avg_time << " ms/image\n";
    std::cout << "   Throughput:       " << std::fixed << std::setprecision(1)
              << (1000.0 / avg_time) << " images/sec\n";
}

void printPerClassAccuracy(const Statistics& stats) {
    std::cout << "\n📈 Per-Class Accuracy:\n";
    std::cout << "   " << std::string(60, '─') << "\n";
    std::cout << "   " << std::left << std::setw(8) << "Digit"
              << std::setw(20) << "Correct/Total"
              << std::setw(15) << "Accuracy"
              << "Bar\n";
    std::cout << "   " << std::string(60, '─') << "\n";

    for (int digit = 0; digit < 10; ++digit) {
        int class_total = stats.per_class_total.at(digit);
        int class_correct = stats.per_class_correct.at(digit);

        if (class_total > 0) {
            double acc = stats.getClassAccuracy(digit);
            int bar_len = static_cast<int>(acc / 2); // Scale to 50 chars max

            std::cout << "   " << std::left << std::setw(8) << digit
                      << std::right << std::setw(5) << class_correct << "/"
                      << std::left << std::setw(5) << class_total
                      << std::string(8, ' ')
                      << std::right << std::setw(6) << std::fixed
                      << std::setprecision(2) << acc << "%"
                      << std::string(8, ' ') << "[";

            for (int i = 0; i < 50; ++i) {
                if (i < bar_len) std::cout << "█";
                else std::cout << "░";
            }
            std::cout << "]\n";
        }
    }
}

void printConfusionMatrix(const Statistics& stats) {
    std::cout << "\n🔄 Confusion Matrix:\n";
    std::cout << "   (Rows: True labels, Columns: Predicted labels)\n";
    std::cout << "   " << std::string(65, '─') << "\n";

    // Header
    std::cout << "      ";
    for (int i = 0; i < 10; ++i) {
        std::cout << std::setw(5) << i << " ";
    }
    std::cout << "\n   " << std::string(65, '─') << "\n";

    // Matrix rows
    for (int i = 0; i < 10; ++i) {
        std::cout << "   " << std::setw(2) << i << " ";
        for (int j = 0; j < 10; ++j) {
            int count = stats.confusion_matrix[i][j];

            if (i == j && count > 0) {
                // Diagonal (correct predictions) - green
                std::cout << GREEN << std::setw(5) << count << RESET << " ";
            } else if (count > 0) {
                // Errors - red
                std::cout << RED << std::setw(5) << count << RESET << " ";
            } else {
                std::cout << std::setw(5) << count << " ";
            }
        }
        std::cout << "\n";
    }
}

void printErrorAnalysis(const Statistics& stats) {
    std::cout << "\n❌ Error Analysis:\n";
    std::cout << "   " << std::string(60, '─') << "\n";

    int total_errors = stats.total - stats.correct;
    if (total_errors == 0) {
        std::cout << "   " << GREEN << "No errors! Perfect accuracy!" << RESET << "\n";
        return;
    }

    std::cout << "   Total errors: " << total_errors << "\n\n";

    for (int digit = 0; digit < 10; ++digit) {
        int class_total = stats.per_class_total.at(digit);
        int class_correct = stats.per_class_correct.at(digit);
        int errors = class_total - class_correct;

        if (errors > 0) {
            std::cout << "   Digit " << digit << ": " << errors << " error(s)\n";

            // Find top misclassifications
            std::vector<std::pair<int, int>> misclass; // (predicted_digit, count)
            for (int j = 0; j < 10; ++j) {
                if (j != digit && stats.confusion_matrix[digit][j] > 0) {
                    misclass.push_back({j, stats.confusion_matrix[digit][j]});
                }
            }

            // Sort by count (descending)
            std::sort(misclass.begin(), misclass.end(),
                      [](const auto& a, const auto& b) { return a.second > b.second; });

            // Show top 3 misclassifications
            int show_count = std::min(3, static_cast<int>(misclass.size()));
            for (int i = 0; i < show_count; ++i) {
                std::cout << "      → Often misclassified as " << misclass[i].first
                          << ": " << misclass[i].second << " times\n";
            }
        }
    }
}

void processBatchInference(const std::string& model_path, const std::string& test_dir) {
    std::cout << "Loading model from: " << model_path << "\n";
    std::cout << std::string(80, '=') << "\n";

    // Initialize inference engine
    MNISTInference inference;
    if (!inference.loadModel(model_path)) {
        std::cerr << RED << "✗ Failed to load model" << RESET << "\n";
        return;
    }

    std::cout << "\nProcessing test images from: " << test_dir << "\n";
    std::cout << std::string(80, '=') << "\n\n";

    Statistics stats;

    // Process each digit folder (0-9)
    for (int digit = 0; digit < 10; ++digit) {
        fs::path digit_dir = fs::path(test_dir) / std::to_string(digit);

        if (!fs::exists(digit_dir) || !fs::is_directory(digit_dir)) {
            std::cout << YELLOW << "⚠️  Warning: Directory " << digit_dir
                      << " not found, skipping..." << RESET << "\n";
            continue;
        }

        // Collect all image files
        std::vector<fs::path> image_files;
        for (const auto& entry : fs::directory_iterator(digit_dir)) {
            if (entry.is_regular_file()) {
                std::string ext = entry.path().extension().string();
                std::transform(ext.begin(), ext.end(), ext.begin(), ::tolower);

                if (ext == ".png" || ext == ".jpg" || ext == ".jpeg" || ext == ".bmp") {
                    image_files.push_back(entry.path());
                }
            }
        }

        if (image_files.empty()) {
            std::cout << YELLOW << "⚠️  Warning: No images found in " << digit_dir
                      << ", skipping..." << RESET << "\n";
            continue;
        }

        // Process images
        for (size_t i = 0; i < image_files.size(); ++i) {
            try {
                auto start_time = std::chrono::high_resolution_clock::now();

                MNISTInference::Result result;
                if (inference.predict(image_files[i].string(), result)) {
                    auto end_time = std::chrono::high_resolution_clock::now();
                    std::chrono::duration<double> elapsed = end_time - start_time;

                    stats.addPrediction(digit, result.predicted_digit, elapsed.count());
                } else {
                    std::cerr << "\n" << RED << "⚠️  Error processing: "
                              << image_files[i].filename() << RESET << "\n";
                }

                // Update progress bar every 50 images
                if ((i + 1) % 50 == 0 || (i + 1) == image_files.size()) {
                    printProgressBar(i + 1, image_files.size(), digit);
                }

            } catch (const std::exception& e) {
                std::cerr << "\n" << RED << "⚠️  Exception: " << e.what() << RESET << "\n";
            }
        }

        // Print summary for this digit
        int class_total = stats.per_class_total[digit];
        int class_correct = stats.per_class_correct[digit];
        double class_acc = stats.getClassAccuracy(digit);

        std::cout << "\r  Digit " << digit << ": " << class_correct << "/"
                  << class_total << " (" << std::fixed << std::setprecision(2)
                  << class_acc << "%) " << GREEN << "✓" << RESET << "\n";
    }

    std::cout << "\n" << std::string(80, '=') << "\n";

    // Print detailed report
    printHeader();
    printOverallStats(stats);
    printPerformance(stats);
    printPerClassAccuracy(stats);
    printConfusionMatrix(stats);
    printErrorAnalysis(stats);

    std::cout << "\n" << std::string(80, '=') << "\n";
    std::cout << GREEN << "\n✅ Batch inference complete!\n" << RESET << "\n";
}

int main(int argc, char* argv[]) {
    std::cout << "\n╔" << std::string(60, '=') << "╗\n";
    std::cout << "║" << std::string(10, ' ') << "MNIST Batch Inference - C++" << std::string(22, ' ') << "║\n";
    std::cout << "╚" << std::string(60, '=') << "╝\n\n";

    // Parse command line arguments
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " <model_path> <test_dir>\n\n";
        std::cerr << "Arguments:\n";
        std::cerr << "  model_path  Path to TorchScript model (.pt file)\n";
        std::cerr << "  test_dir    Root directory containing digit folders (0-9)\n\n";
        std::cerr << "Example:\n";
        std::cerr << "  " << argv[0] << " tinycnn_mnist_traced.pt ./tests\n\n";
        return 1;
    }

    std::string model_path = argv[1];
    std::string test_dir = argv[2];

    std::cout << "Configuration:\n";
    std::cout << "  Model:     " << model_path << "\n";
    std::cout << "  Test dir:  " << test_dir << "\n\n";

    try {
        processBatchInference(model_path, test_dir);
    } catch (const std::exception& e) {
        std::cerr << RED << "\n✗ Fatal error: " << e.what() << RESET << "\n";
        return 1;
    }

    return 0;
}

File: batch_inference.py
#!/usr/bin/env python3
"""
Batch inference on categorized test images.

This script performs inference on images organized in digit-labeled folders
(tests/0/, tests/1/, ..., tests/9/) and reports per-class accuracy.
"""

import torch
import torch.nn.functional as F
from PIL import Image
import numpy as np
import os
from pathlib import Path
from collections import defaultdict
import time
from model import TinyCNN


class BatchInference:
    """Batch inference engine with statistics tracking."""

    def __init__(self, model_path: str):
        """
        Initialize batch inference engine.

        Args:
            model_path: Path to model file (.pt)
        """
        self.model_path = model_path
        self.model = None
        self.results = {
            'total': 0,
            'correct': 0,
            'per_class': defaultdict(lambda: {'total': 0, 'correct': 0}),
            'confusion_matrix': np.zeros((10, 10), dtype=int),
            'errors': [],
            'inference_times': []
        }

    def load_model(self):
        """Load PyTorch model."""
        print(f"Loading model from: {self.model_path}")

        checkpoint = torch.load(self.model_path, map_location=torch.device('cpu'))

        if isinstance(checkpoint, dict):
            print("  ✓ Detected state_dict format")
            self.model = TinyCNN()
            self.model.load_state_dict(checkpoint)
        else:
            self.model = checkpoint

        self.model.eval()
        print("  ✓ Model loaded and ready for inference\n")

    def preprocess_image(self, image_path: str) -> torch.Tensor:
        """
        Preprocess image for model input.

        Args:
            image_path: Path to image file

        Returns:
            Preprocessed tensor [1, 1, 28, 28]
        """
        # Load image
        img = Image.open(image_path).convert('L')  # Convert to grayscale

        # Resize to 28x28 if needed
        if img.size != (28, 28):
            img = img.resize((28, 28), Image.LANCZOS)

        # Convert to numpy array and normalize
        img_array = np.array(img, dtype=np.float32) / 255.0

        # Apply MNIST normalization
        #mean = 0.1307
        #std = 0.3081
        mean = 0.5
        std = 0.5
        img_array = (img_array - mean) / std

        # Convert to tensor
        tensor = torch.from_numpy(img_array).unsqueeze(0).unsqueeze(0)  # [1, 1, 28, 28]

        return tensor

    def predict(self, image_path: str) -> tuple:
        """
        Run inference on a single image.

        Args:
            image_path: Path to image file

        Returns:
            Tuple of (predicted_digit, confidence, inference_time)
        """
        start_time = time.time()

        # Preprocess
        input_tensor = self.preprocess_image(image_path)

        # Inference
        with torch.no_grad():
            output = self.model(input_tensor)
            probabilities = F.softmax(output, dim=1)
            confidence, predicted = torch.max(probabilities, 1)

        inference_time = time.time() - start_time

        return predicted.item(), confidence.item(), inference_time

    def process_directory(self, test_dir: str, show_progress: bool = True):
        """
        Process all images in categorized directories.

        Args:
            test_dir: Root directory containing digit folders (0-9)
            show_progress: Show progress during processing
        """
        test_path = Path(test_dir)

        if not test_path.exists():
            raise FileNotFoundError(f"Test directory not found: {test_dir}")

        print(f"Processing test images from: {test_dir}")
        print("=" * 80)

        # Process each digit folder (0-9)
        for digit in range(10):
            digit_dir = test_path / str(digit)

            if not digit_dir.exists():
                print(f"⚠️  Warning: Directory {digit_dir} not found, skipping...")
                continue

            # Get all image files
            image_files = []
            for ext in ['*.png', '*.jpg', '*.jpeg', '*.bmp']:
                image_files.extend(digit_dir.glob(ext))

            if not image_files:
                print(f"⚠️  Warning: No images found in {digit_dir}, skipping...")
                continue

            # Process images
            for idx, img_path in enumerate(image_files):
                try:
                    predicted, confidence, inf_time = self.predict(str(img_path))

                    # Update statistics
                    self.results['total'] += 1
                    self.results['per_class'][digit]['total'] += 1
                    self.results['confusion_matrix'][digit][predicted] += 1
                    self.results['inference_times'].append(inf_time)

                    if predicted == digit:
                        self.results['correct'] += 1
                        self.results['per_class'][digit]['correct'] += 1
                    else:
                        # Record error
                        self.results['errors'].append({
                            'image': str(img_path),
                            'true_label': digit,
                            'predicted': predicted,
                            'confidence': confidence
                        })

                    # Show progress
                    if show_progress and (idx + 1) % 50 == 0:
                        print(f"  Digit {digit}: Processed {idx + 1}/{len(image_files)} images", end='\r')

                except Exception as e:
                    print(f"\n⚠️  Error processing {img_path}: {e}")

            if show_progress:
                total_digit = self.results['per_class'][digit]['total']
                correct_digit = self.results['per_class'][digit]['correct']
                accuracy_digit = 100.0 * correct_digit / total_digit if total_digit > 0 else 0
                print(f"  Digit {digit}: {correct_digit}/{total_digit} "
                      f"({accuracy_digit:.2f}%) ✓")

        print("\n" + "=" * 80)

    def print_report(self):
        """Print detailed inference report."""
        print("\n" + "╔" + "=" * 78 + "╗")
        print("║" + " " * 25 + "BATCH INFERENCE REPORT" + " " * 31 + "║")
        print("╚" + "=" * 78 + "╝\n")

        # Overall accuracy
        overall_acc = 100.0 * self.results['correct'] / self.results['total']
        print(f"📊 Overall Statistics:")
        print(f"   Total images:     {self.results['total']:,}")
        print(f"   Correct:          {self.results['correct']:,}")
        print(f"   Incorrect:        {self.results['total'] - self.results['correct']:,}")
        print(f"   Accuracy:         {overall_acc:.2f}%")

        # Performance
        avg_time = np.mean(self.results['inference_times']) * 1000
        print(f"\n⚡ Performance:")
        print(f"   Avg inference:    {avg_time:.2f} ms/image")
        print(f"   Throughput:       {1000.0 / avg_time:.1f} images/sec")

        # Per-class accuracy
        print(f"\n📈 Per-Class Accuracy:")
        print("   " + "─" * 60)
        print(f"   {'Digit':<8} {'Correct/Total':<20} {'Accuracy':<15} {'Bar'}")
        print("   " + "─" * 60)

        for digit in range(10):
            stats = self.results['per_class'][digit]
            if stats['total'] > 0:
                acc = 100.0 * stats['correct'] / stats['total']
                bar_len = int(acc / 2)  # Scale to 50 chars max
                bar = "█" * bar_len + "░" * (50 - bar_len)
                print(f"   {digit:<8} {stats['correct']:>5}/{stats['total']:<5}        "
                      f"{acc:>6.2f}%        [{bar}]")
            else:
                print(f"   {digit:<8} {'N/A':<20} {'N/A':<15}")

        # Confusion Matrix
        print(f"\n🔄 Confusion Matrix:")
        print("   (Rows: True labels, Columns: Predicted labels)")
        print("   " + "─" * 65)
        print("      ", end="")
        for i in range(10):
            print(f"{i:>5}", end=" ")
        print()
        print("   " + "─" * 65)

        for i in range(10):
            print(f"   {i:>2} ", end="")
            for j in range(10):
                count = self.results['confusion_matrix'][i][j]
                if i == j:
                    # Diagonal (correct predictions) - highlighted
                    print(f"\033[1;32m{count:>5}\033[0m", end=" ")
                elif count > 0:
                    # Errors - highlighted in red
                    print(f"\033[1;31m{count:>5}\033[0m", end=" ")
                else:
                    print(f"{count:>5}", end=" ")
            print()

        # Error analysis
        if self.results['errors']:
            print(f"\n❌ Error Analysis (Total: {len(self.results['errors'])}):")
            print("   " + "─" * 60)

            # Group errors by true label
            errors_by_class = defaultdict(list)
            for error in self.results['errors']:
                errors_by_class[error['true_label']].append(error)

            for digit in range(10):
                if digit in errors_by_class:
                    errors = errors_by_class[digit]
                    # Count most common misclassifications
                    misclass_counts = defaultdict(int)
                    for err in errors:
                        misclass_counts[err['predicted']] += 1

                    top_misclass = sorted(misclass_counts.items(),
                                          key=lambda x: x[1], reverse=True)[:3]

                    print(f"   Digit {digit}: {len(errors)} errors")
                    for pred_digit, count in top_misclass:
                        print(f"      → Often misclassified as {pred_digit}: {count} times")

            # Show some error examples
            print(f"\n   Top 5 Error Examples:")
            sorted_errors = sorted(self.results['errors'],
                                   key=lambda x: x['confidence'], reverse=True)[:5]

            for i, error in enumerate(sorted_errors, 1):
                print(f"   {i}. {Path(error['image']).name}")
                print(f"      True: {error['true_label']}, "
                      f"Predicted: {error['predicted']} "
                      f"(confidence: {error['confidence'] * 100:.2f}%)")

        print("\n" + "=" * 80)

    def save_results(self, output_path: str = "inference_results.csv"):
        """
        Save detailed results to CSV.

        Args:
            output_path: Path to output CSV file
        """
        import csv

        with open(output_path, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(['Image', 'True Label', 'Predicted', 'Confidence', 'Correct'])

            # Write all predictions
            for digit in range(10):
                digit_dir = Path(f"tests/{digit}")
                if digit_dir.exists():
                    for img_path in digit_dir.glob("*.png"):
                        predicted, confidence, _ = self.predict(str(img_path))
                        correct = 'Yes' if predicted == digit else 'No'
                        writer.writerow([str(img_path), digit, predicted,
                                         f"{confidence:.4f}", correct])

        print(f"✓ Results saved to: {output_path}")


def main():
    import argparse

    parser = argparse.ArgumentParser(description='Batch inference on categorized test images')
    parser.add_argument('--model', type=str, default='best_model_mnist_and_water_meter.pt',
                        help='Path to model file (.pt)')
    parser.add_argument('--test-dir', type=str, default='./tests',
                        help='Root directory containing digit folders (0-9)')
    parser.add_argument('--save-csv', action='store_true',
                        help='Save detailed results to CSV')
    parser.add_argument('--csv-output', type=str, default='inference_results.csv',
                        help='Output CSV filename')

    args = parser.parse_args()

    # Initialize batch inference
    batch_inf = BatchInference(args.model)

    # Load model
    batch_inf.load_model()

    # Process all images
    batch_inf.process_directory(args.test_dir)

    # Print report
    batch_inf.print_report()

    # Save results if requested
    if args.save_csv:
        print(f"\nSaving detailed results to CSV...")
        batch_inf.save_results(args.csv_output)

    print(f"\n✅ Batch inference complete!\n")


if __name__ == "__main__":
    main()
